### ISSUE TITLE
Miscellaneous Remarks

### ISSUE THREAD

**Issue by matthieu-m** [CAD97/storages-api#2]
https://github.com/CAD97/storages-api/blob/3f43e2f0e21036868b9c5b42ec7b6bee2e6a9f47/src/traits.rs#L25

I notice that in a departure from [`storage-poc`](https://github.com/matthieu-m/storage-poc) the `Storage` trait here takes a strongly typed parameter.

This is the direction storage-poc had started with, and it switched for a number of reasons, the main one being encapsulation. I note that you haven't yet implemented a linked list or binary tree, when you do you'll need a storage which stores a "Node" which is an implementation detail: the user supplying the storage doesn't know the Node type by which the storage should be parameterized.

There are other reasons, such as flexibility, as in the ability for a collection to use the same storage object for various types of allocations, which seems fairly exotic, but encapsulation is the big one, I think.

https://github.com/CAD97/storages-api/blob/3f43e2f0e21036868b9c5b42ec7b6bee2e6a9f47/src/traits.rs#L23

I started without `allocate`/`deallocate`, but added them back in https://github.com/matthieu-m/storage-poc/commit/a84ca554658a4b057da5868e21c525e4fb1bd4f8 .

The problem is that Rust does not yet allow passing DST by value (AFAIK), and therefore if the user wishes to move an already unsized value into an allocation, they cannot use `create`.

While not a common usecase, it's simple enough to support it, so I'd advise to.

**Comment by CAD97** [comment]
I originally had `trait Storage` with a GAT `type Handle<T>`, but I switched to `trait Storage<T>` because of #1.

Specifically, #1 is about emulating `&move T` via `Box<T, RefStorage<&mut ManuallyDrop<T>>>`. [`RefStorage`](https://cad97.github.io/storages-api/storage_api/struct.RefStorage.html) is quite awkward to support, because of the major restriction that it *must* leave a valid `T` behind. At this point, I don't know if this actually fits the `Storage` shape, since it has the extra restriction of `create` always failing, as when constructed, it already has a live singleton handle that can't be destroyed.

A goal of the storage API in my eyes is that it should be able to provide "`dyn* Trait`"; that is, a single type which can hold all of the following while remaining just 2 `usize` big:

- `&mut dyn Trait`
- `&move dyn Trait` (maybe)
- `Box<dyn Trait, AllocStorage<Global>>`
- `Box<dyn Trait, InlineStorage<usize>>`

How to do this is discussed in abstract by [nikomatsakis](https://smallcultfollowing.com/babysteps//blog/2022/03/29/dyn-can-we-make-dyn-sized/) and [myself](https://dev.to/cad97/dyn-doesnt-need-to-be-special-3ldm), so it's a matter of actually implementing it at this point. (TL;DR: `&move dyn Trait`, wrap the vtable `drop_in_place` to do the right thing. Inline storage wraps the rest of the vtable as well to pass the small object as `ptr::invalid`.)

-----

I wish Rust + GAT would allow us to directly write `S: for<T> Storage<T>`, but that's a bit much maybe. The storage API is already very complexly generic. I'm currently leaning towards reverting `RefStorage`. Losing it doesn't completely lose `&move T` style storages, just the `&move ?Sized` constructor for them. `BorrowedStorage` (not yet implemented) works on `&mut MaybeUninit<DataStore>`, supports `create`/`destroy`, and doesn't have the `T` validity restriction. The only limitation is that it has to be created from a sized place.

Another benefit of dropping `RefStorage` is that all handles uniformly contain the pointee metadata again. I've been asked to also check what the storage API looks like dealing in untyped `Layout` allocation rather than typed metadata allocation[^1], and such would store `(UntypedHandle, Metadata)` as a handle, and just not support `RefStorage` at all.

[^1]: There is a notable benefit to this. `Layout` enforces that an allocation of the layout is a valid Rust allocation — notably, that size is <= `usize::MAX`, and we're looking into making it enforce that size is <= `isize::MAX` to make it actually guaranteed, and take that nebulous requirement out of being enforced by practice (that std alloc has gotten wrong multiple times). With metadata, however, there's no such restriction; you can freely and safely construct slice metadata describing slices so long that size calculation (`size_of::<T>() * len`) overflows back to a reasonable size < `isize::MAX`. `size_of_val_raw` is unsafe, requiring that the size calculation does not exceed `isize::MAX`; there's no way to generically check that a metadata describes a valid layout, even after constructing it. I've an experiment open to see the cost of just always saturating the size calculation, but avoiding the problem altogether by using `Layout` seems desirable.

-----

Note also that my `create`/`destroy` are `allocate`/`deallocate`, in that they take a describer of the place, rather than the object itself. I'll probably just end up renaming them to `allocate`/`deallocate`.

**Comment by CAD97** [comment]
I reverted the typed `Storage` change; I agree with you on the benefits of storage not having a strongly typed parameter outweigh the (small) benefits.

I've a new new revision over at https://github.com/CAD97/storages-api/pull/3 that I'd like your opinion on if you have time; at the request of Thom I've experimented with implementing the Storage API on `Layout` directly rather than metadata, and I think it ultimately provides for a simplification and more gradual complexity.

I think the remarks contained in this issue are resolved.
